### PR TITLE
actions: cross build in ci and fix publish

### PR DESCRIPTION
**Pull request by jan--f**
This add

- a job to build on 3 architectures in the CI action
- Fixes the publish action to cross build all architectures first

~TODO: add a release action and discuss and add appropriate trigger filters for release and publish.~

### CODE REVIEW

File: .github/workflows/publish.yml [hunk 1]
```diff
@@ -2,23 +2,38 @@
...
+          thread: ${{ matrix.thread }}
   publish_main:
     name: Publish main branch artefacts
     runs-on: ubuntu-latest
-    needs: ci
+    needs: build
     steps:
       - uses: actions/checkout@v3
-      - uses: prometheus/promci@v0.0.2
+      - uses: prometheus/promci@3cb0c3871f223bd5ce1226995bd52ffb314798b6 # v0.1.0
```
Comment (gotjosh) [inline .github/workflows/publish.yml:32]: Why are we downgrading here? That's wrong with the latest version?

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -2,23 +2,38 @@
...
       - uses: ./.github/promci/actions/publish_main
         with:
           docker_hub_login: ${{ secrets.docker_hub_login }}
           docker_hub_password: ${{ secrets.docker_hub_password }}
+          quay_io_organization: ${{ secrets.quay_io_organization }}
           quay_io_login: ${{ secrets.quay_io_login }}
           quay_io_password: ${{ secrets.quay_io_password }}
```
Comment (gotjosh) [inline .github/workflows/publish.yml:39]: None of these secrets are defined. At which point would be ready to test them? How did you approach this in Prometheus?

**Comment by gotjosh** [comment]
Looking at the current failures, I also see:

> The `set-output` command is deprecated and will be disabled soon. Please upgrade to using Environment Files. For more information see: https://github.blog/changelog/2022-10-11-github-actions-deprecating-save-state-and-set-output-commands/

File: .github/workflows/publish.yml (hunk 1 above)
Comment (jan--f) [reply .github/workflows/publish.yml:32]: This is actually an upgrade from `v0.0.2` to `v0.1.0` :) The pinning to a sha I copied from Prometheus, seems like good practice and I intend to do that for the other instances too.

File: .github/workflows/publish.yml (hunk 2 above)
Comment (jan--f) [reply .github/workflows/publish.yml:39]: Yeah I'm looking into that, not sure myself at this point. In my clone I defined those secrets but it still fails, there is something I don't understand with multi arch builds.

[If those secrets don't exist the publish action is just a no-op](https://github.com/prometheus/promci/blob/3cb0c3871f223bd5ce1226995bd52ffb314798b6/actions/publish_main/action.yml#L31-L38). so we can leave this for now and once we understand what's going on, defining the secrets is all that is needed to publish from the action.

Comment (gotjosh) [reply .github/workflows/publish.yml:39]: Great to know!

File: .github/workflows/publish.yml (hunk 1 above)
Comment (gotjosh) [reply .github/workflows/publish.yml:32]: 🤦 I need to learn to read a bit better - agreed, pinning of versions saves us from unexpected failures.

File: .github/workflows/publish.yml (hunk 2 above)
Comment (roidelapluie) [reply .github/workflows/publish.yml:39]: they should be defined at the prometheus org level

File: .github/workflows/publish.yml [hunk 3]
```diff
@@ -2,23 +2,38 @@
 name: Publish
 on:  # yamllint disable-line rule:truthy
   push:
-    branches:
-      - main
+    # branches:
+    #   - main
```
Comment (roidelapluie) [inline .github/workflows/publish.yml:6]: why?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -16,6 +16,21 @@ jobs:
       - run: make apiv2
       - run: git diff --exit-code
 
+  build:
+    name: Build Alertmanager for common architectures
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        thread: [ 0, 1, 2 ]
```
Comment (roidelapluie) [inline .github/workflows/ci.yml:24]: This complexity is not needed for alertmanager

File: .github/workflows/publish.yml (hunk 3 above)
Comment (jan--f) [reply .github/workflows/publish.yml:6]: This is just temporary to tigger the action on a PR push. I'll drop all `DEBUG` commits before this merges.

File: .github/workflows/ci.yml (hunk 4 above)
Comment (jan--f) [reply .github/workflows/ci.yml:24]: I took this from the[ circleci build job](https://github.com/prometheus/alertmanager/blob/c6be0bcabf37e837d24786e3d5c10565ac499531/.circleci/config.yml#L127-L135) thinking its a reasonable check before merge. Happy to drop if it isn't needed.

Comment (jan--f) [reply .github/workflows/ci.yml:24]: Also https://github.com/prometheus/alertmanager/pull/3530 wants to extend this.

File: .github/workflows/publish.yml [hunk 5]
```diff
@@ -2,23 +2,38 @@
...
+    needs: build
     steps:
       - uses: actions/checkout@v3
-      - uses: prometheus/promci@v0.0.2
+      - uses: prometheus/promci@3cb0c3871f223bd5ce1226995bd52ffb314798b6 # v0.1.0
       - uses: ./.github/promci/actions/publish_main
         with:
           docker_hub_login: ${{ secrets.docker_hub_login }}
           docker_hub_password: ${{ secrets.docker_hub_password }}
+          quay_io_organization: ${{ secrets.quay_io_organization }}
```
Comment (gotjosh) [inline .github/workflows/publish.yml:37]: @roidelapluie is right these all are defined at an org-level except this one - why do we need this one again?

Comment (jan--f) [reply .github/workflows/publish.yml:37]: Again a debug commit to change the org to my quay org.

**Comment by jan--f** [comment]
@gotjosh sorry, let me drop the debug commits to reduce noise.

**Comment by jan--f** [comment]
Ok I added the build step before the release action as well. 
@roidelapluie @gotjosh Is there anything else needed here?

**Review by gotjosh — APPROVED** [review]

**Pull request merged**